src/libraries/UniswapV3Lib.sol

```diff
@@ -0,0 +1,259 @@
+    /*** Helper functions                                                                       ***/
+    /**********************************************************************************************/
+    
+    //-- Swap helper functions
+    function _populateSwapCache(UniV3Context calldata context, SwapParams calldata params) internal view returns (SwapCache memory) {
+        IUniswapV3PoolLike pool         = IUniswapV3PoolLike(context.pool);
+        address token0                  = pool.token0();
+        address token1                  = pool.token1();
+        (, int24 currentTick, , , , , ) = pool.slot0();
+
+        int24 delta = int24(params.tickDelta);
+
+        address tokenOut = params.tokenIn == token0 ? token1 : token0;
+        
+        // Expected out is calculated by by converting amountIn to amountOut using the TWAP tick since some time ago
+        (int24 twapExpectedOutTick, ) = UniswapV3OracleLibrary.consult(context.pool, params.poolParams.swapTwapSecondsAgo); 
+        uint256 twapExpectedOut       = UniswapV3OracleLibrary.getQuoteAtTick(twapExpectedOutTick, uint128(params.amountIn), params.tokenIn, tokenOut);
+
+        int24 limitTick;
+        if (params.tokenIn == token0) {
+            limitTick = currentTick - delta;
+            if (limitTick < TickMath.MIN_TICK) limitTick = TickMath.MIN_TICK;
+        } else {
+            limitTick = currentTick + delta;
+            if (limitTick > TickMath.MAX_TICK) limitTick = TickMath.MAX_TICK;
+        }
+
+        uint160 sqrtPriceLimitX96 = TickMath.getSqrtRatioAtTick(limitTick);
+
+        return SwapCache({
+            pool             : pool,
+            token0           : token0,
+            token1           : token1,
+            tokenOut         : tokenOut,
+            sqrtPriceLimitX96: sqrtPriceLimitX96,
+            twapExpectedOut  : twapExpectedOut
+        });
+    }
+
+    function _callSwap(UniV3Context calldata context, SwapParams calldata params, SwapCache memory cache) internal returns (uint256 amountOut) {
+        address tokenOut = params.tokenIn == cache.token0 ? cache.token1 : cache.token0;
+
+        bytes memory result = context.proxy.doCall(
+            params.router,
+            abi.encodeWithSelector(
+                ISwapRouter.exactInputSingle.selector,
+                ISwapRouter.ExactInputSingleParams({
+                    tokenIn          : params.tokenIn,
+                    tokenOut         : tokenOut,
+                    fee              : cache.pool.fee(),
+                    recipient        : address(context.proxy),
+                    deadline         : context.deadline,
+                    amountIn         : params.amountIn,
+                    amountOutMinimum : params.minAmountOut,
+                    sqrtPriceLimitX96: cache.sqrtPriceLimitX96
+                })
+            )
+        );
+
+        amountOut = abi.decode(result, (uint256));
+    }
+
+    //-- General helper functions
+    function _approve(
```

Comment:
Not for this PR, but looking at the other libraries, it seems like we have a proof for basically all of them. We should probably move this to a common library.

Reply:
https://github.com/grove-labs/grove-alm-controller/pull/56

Reply:
https://github.com/grove-labs/grove-alm-controller/pull/54/commits/3c110e5baceb11c885d49655d75eebfe2f90299d